drivers/adc/adc_mcp356x.c

```diff
@@ -0,0 +1,836 @@
+ * get an accurate sample rate an external MCLK is needed.  Using the ADC's timer means the time
+ * between samples will not be affected by Zephyr interrupt latency or thread scheduling, and
+ * should have much less jitter than using a Zephyr timer.
+ *
+ * Using the ADC timer limits the delay to 2^24 ticks of DMCLK (~13.65 sec with prescale 1 and
+ * default MCLK).  interval_us will be the time between the last channel of one sample set and
+ * the first channel of the next.  I.e., it does NOT include the time to sample the channels.
+ * Sampling will begin immediately on an adc_read() call and not wait interval_us first.
+ *
+ * In ADC timer mode, channel 16 will run in contiuous mode, which allows for a higher sample
+ * rate.  Channel 16 does not support non-zero interval_us values in ADC timer mode.
+ *
+ * Alternatively, a Zephyr timer can be used to generate the sampling rate.  This requires a SPI
+ * transfer to start each conversion, and this will add jitter due to Zephyr interrupt latency
+ * and thread scheduling delays.  interval_us will be the time between starts of each channel
+ * set.  I.e., it does include the sampling time.  If the time to sample the channels is greater
+ * than interval_us, the one-shot samples effectively run back-to-back, possibly slower than
+ * interval_us.  The first sample from an adc_read() call does not start until after a
+ * interval_us delay.  Channel 16 does not run in continuous mode and can not sample at as high
+ * of a rate.  This is not just because of extra SPI commands and more overhead.  The manner in
+ * which a ΔΣ ADC creates the samples is different between continuous conversion and a sequence
+ * of one-shot conversions.
+ *
+ * Setting calibrate in the adc_sequence will enable the offset cancellatin algorithm, see
+ * §5.1.3, which causes the ADC to sample twice, with the mux set in either direction, to
+ * attempt to cancel out offset error.  This doubles the time each sample takes.
+ */
+
+/* Missing features.
+ *
+ * Configure bias current (in device tree?).
+ * Configure internal clock output to MCLK pin (dts).
+ * Standby vs shutdown when not sampling (dts?).
+ * Tweak IRQ timeout to just to expected sampling rate.
+ * Support the CRC data checking mode.
+ * Detect POR or unexpected config change via status bits when reading data.
+ * Timestamp the DRDY interrupt, to get more accurate data timestamps.
+ */
+
+#include <drivers/adc.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <kernel.h>
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+#include <zephyr.h>
+
+#include "adc_mcp356x.h"
+
+LOG_MODULE_REGISTER(adc_mcp356x, CONFIG_ADC_LOG_LEVEL);
+
+/* Verbose debug logging.  Individual register access. IRQ timing. */
+#define VERBOSE_LOG	0
+
+#if VERBOSE_LOG
+#define LOG_VDBG(...)	LOG_DBG(__VA_ARGS__)
+#else
+#define LOG_VDBG(...)	(void)(0)
+#endif
```

Comment:
We don't usually do this - is this really needed after the driver was developed?